File: cef-web/src/main/java/gov/epa/cef/web/service/ReportingPeriodService.java
package gov.epa.cef.web.service;

import java.util.List;

import gov.epa.cef.web.service.dto.EmissionBulkEntryHolderDto;
import gov.epa.cef.web.service.dto.ReportingPeriodBulkEntryDto;
import gov.epa.cef.web.service.dto.ReportingPeriodDto;
import gov.epa.cef.web.service.dto.ReportingPeriodUpdateResponseDto;

public interface ReportingPeriodService {

    /**
     * Create a new Reporting Period
     * @param dto
     * @return
     */
    public ReportingPeriodDto create(ReportingPeriodDto dto);

    /**
     * Update a Reporting Period
     * @param dto
     * @return
     */
    public ReportingPeriodUpdateResponseDto update(ReportingPeriodDto dto);

    /**
     * Retrieve Reporting Period by id
     * @param id
     * @return
     */
    ReportingPeriodDto retrieveById(Long id);

    /**
     * Retrieve Reporting Periods for an emissions process
     * @param processId
     * @return
     */
    List<ReportingPeriodDto> retrieveForEmissionsProcess(Long processId);

    /**
     * Retrieve Reporting Periods for Bulk Entry for a specific facility site
     * @param facilitySiteId
     * @return
     */
    public List<ReportingPeriodBulkEntryDto> retrieveBulkEntryReportingPeriodsForFacilitySite(Long facilitySiteId);

    /**
     * Update the throughput for multiple Reporting Periods at once
     * @param facilitySiteId
     * @param dtos
     * @return
     */
    public List<EmissionBulkEntryHolderDto> bulkUpdate(Long facilitySiteId, List<ReportingPeriodBulkEntryDto> dtos);

}

File: cef-web/src/main/webapp/src/app/core/services/reporting-period.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ReportingPeriod } from 'src/app/shared/models/reporting-period';
import { ReportingPeriodUpdateResult } from 'src/app/shared/models/reporting-period-update-result';
import { BulkEntryReportingPeriod } from 'src/app/shared/models/bulk-entry-reporting-period';

@Injectable({
  providedIn: 'root'
})
export class ReportingPeriodService {

  private baseUrl = 'api/reportingPeriod';  // URL to web api

  constructor(private http: HttpClient) { }

  create(period: ReportingPeriod): Observable<ReportingPeriod> {
    const url = `${this.baseUrl}`;
    return this.http.post<ReportingPeriod>(url, period);
  }

  update(period: ReportingPeriod): Observable<ReportingPeriodUpdateResult> {
    const url = `${this.baseUrl}/${period.id}`;
    return this.http.put<ReportingPeriodUpdateResult>(url, period);
  }

  /** GET specified Reporting Period from the server */
  retrieve(id: number): Observable<ReportingPeriod> {
    const url = `${this.baseUrl}/${id}`;
    return this.http.get<ReportingPeriod>(url);
  }

  retrieveForEmissionsProcess(processId: number): Observable<ReportingPeriod[]> {
    const url = `${this.baseUrl}/process/${processId}`;
    return this.http.get<ReportingPeriod[]>(url);
  }

  retrieveForBulkEntry(facilitySiteId: number): Observable<BulkEntryReportingPeriod[]> {
    const url = `${this.baseUrl}/bulkEntry/${facilitySiteId}`;
    return this.http.get<BulkEntryReportingPeriod[]>(url);
  }

  bulkUpdate(periods: BulkEntryReportingPeriod[]): Observable<ReportingPeriodUpdateResult[]> {
    const url = `${this.baseUrl}/bulkEntry`;
    return this.http.put<ReportingPeriodUpdateResult[]>(url, periods);
  }
}
